regression/pages/common/utils.py

```diff
@@ -0,0 +1,526 @@
+            payload:
+        """
+        response = self.ecommerce_api_client.coupons.post(data=payload)
+        return response['coupon_id']
+
+    def get_coupon_report(self, coupon_id):
+        """
+        Get coupon report using api
+        Args:
+            coupon_id:
+        """
+        return self.ecommerce_api_client.products(coupon_id).get()
+
+    def delete_coupon(self, coupon_id):
+        """
+        Delete the coupon created in the tests and all associated stock
+        records and vouchers.
+        Args:
+            coupon_id:
+        """
+        self.ecommerce_api_client.coupons(coupon_id).delete()
+
+    def get_coupon_codes(self, coupon_id):
+        """
+        Get coupon codes from coupon report
+        Args:
+            coupon_id:
+        """
+        coupon_codes = []
+        items = (
+            self.get_coupon_report(coupon_id)['attribute_values'][0]['value'])
+        for item in items:
+            coupon_codes.append(item['code'])
+        return coupon_codes
+
+
+def get_required_cookies(cookies_dict_list):
+    """
+    Format required cookies
+    Args:
+        cookies_dict_list:
+    """
+    response_dict = {}
+    for cookie in cookies_dict_list:
+        if cookie['name'] == 'ecommerce_csrftoken':
+            response_dict.update({'ecommerce_csrftoken': cookie['value']})
+        elif cookie['name'] == 'ecommerce_sessionid':
+            response_dict.update({'ecommerce_sessionid': cookie['value']})
+    required_cookies = '; '.join('{}={}'.format(
+        key, val) for key, val in response_dict.items())
+    return required_cookies
+
+
+def get_coupon_request(target_url, cookie_str):
+    """
+    Send a Get request and return the response if successful
+    Args:
+      target_url:
+      cookie_str:
+    """
+    kwargs = {}
+    headers = {
+        'Accept': 'text/html,application/xhtml+xml,'
+                  'application/xml;q=0.9,image/webp,*/*;q=0.8',
+        'Accept-Language': 'en-US,en;q=0.8',
+        'Accept-Encoding': 'gzip, deflate, sdch, br',
+        'Cookie': cookie_str
+    }
+    kwargs.update({
+        'headers': headers,
+        'auth': HTTPBasicAuth(AUTH_USERNAME, AUTH_PASSWORD),
+    })
+    response = requests.get(target_url, **kwargs)
+    if response.ok:
+        return response.content
+    else:
+        msg = "Could not use restful API.  Status code: {0}".format(
+            response.status_code)
+        raise ApiException(msg)
+
+
+def get_coupons_from_email(coupon_string):
+    """
+    Read coupon codes and urls from email
+    Ignore first two lines of file as these are headers
+    Args:
+        coupon_string:
+    Returns:
+        coupon codes and urls:
+    """
+    coupons = {}
+    for i, row in enumerate(coupon_string.splitlines()):
+        if i > 3:
+            if row:
+                new_row = row.split(',')
+                coupons.update({new_row[0]: new_row[1]})
+    return coupons
+
+
+def skip_cleanup_if_test_passed():
```

Comment:
Actually when the tests pass the data is cleaned in the test, it's only when it fails that we need to check for unnecessary data. An example is 

For a test where we need to test that a coupon can be used successfuly by 2 users but gives error on 3rd user
A user logs in and enroll in a course using a coupon, before logging out the user un-enrolled from the course.
2nd user logs in and enroll in the course using a coupon, before logging out this user also un-enrolled from the course.
3rd user logs in and couldn't enroll. 

When the test is run completely we don't need to un-enroll any user afterwards. However if there is a break we need to un-enroll the current user if it has enrolled.

The better way of course will be not to un-enroll during the tests but run the cleanup afterwards which un-enroll all users. I am just avoiding this because the tests are already too long and if in the cleanup we log in by multiple users and un-enroll and logout, that will take a lot of time.

I have started using API's in different areas, I will try to perform the un-enrollment using api, that will fix the timing vs antipattern issue